src/app.rs

```diff
@@ -235,15 +241,133 @@ impl App {
         Ok(app)
     }
 
-    async fn initialize_tree(
+    async fn restore_or_initialize_tree(
         database: &Database,
         tree_depth: usize,
         dense_prefix_depth: usize,
         gc_threshold: usize,
         initial_leaf_value: Hash,
+        mmap_file_path: String,
     ) -> AnyhowResult<TreeState> {
-        let mined_items = database.get_commitments_by_status(Status::Mined).await?;
+        let mut mined_items = database.get_commitments_by_status(Status::Mined).await?;
+        mined_items.sort_by_key(|item| item.leaf_index);
+
+        if let Some(tree_state) = Self::get_cached_tree_state(
+            database,
+            tree_depth,
+            dense_prefix_depth,
+            gc_threshold,
+            &initial_leaf_value,
+            &mined_items,
+            &mmap_file_path,
+        )
+        .await?
+        {
+            info!("tree restored from cache");
+            return Ok(tree_state);
+        }
+        let tree_state = Self::initialize_tree(
+            database,
+            tree_depth,
+            dense_prefix_depth,
+            gc_threshold,
+            initial_leaf_value,
+            mined_items,
+            mmap_file_path,
+        )
+        .await?;
+        info!("tree initialization successful");
+        Ok(tree_state)
+    }
+
+    pub fn get_leftover_leaves_and_update_index(
+        index: &mut usize,
+        dense_prefix_depth: usize,
+        mined_items: &Vec<TreeUpdate>,
+    ) -> Vec<ruint::Uint<256, 4>> {
+        let leftover_items = if mined_items.is_empty() {
+            vec![]
+        } else {
+            let max_leaf = mined_items.last().map(|item| item.leaf_index).unwrap();
+            // if the last index is greater then dense_prefix_depth, 1 << dense_prefix_depth
+            // should be the last index in restored tree
+            *index = std::cmp::min(max_leaf, (1 << dense_prefix_depth) - 1);
+
+            if max_leaf - *index == 0 {
+                return vec![];
+            }
+
+            let mut leaves = Vec::with_capacity(max_leaf - *index);
+
+            let leftover = &mined_items[(*index + 1)..];
+
+            for item in leftover {
+                leaves.push(item.element);
+            }
+
+            leaves
+        };
+
+        leftover_items
+    }
+
+    async fn get_cached_tree_state(
+        database: &Database,
+        tree_depth: usize,
+        dense_prefix_depth: usize,
+        gc_threshold: usize,
+        initial_leaf_value: &Hash,
+        mined_items: &Vec<TreeUpdate>,
+        mmap_file_path: &str,
+    ) -> anyhow::Result<Option<TreeState>> {
+        let mut last_mined_index_in_dense: usize = 0;
+        let leftover_items = Self::get_leftover_leaves_and_update_index(
+            &mut last_mined_index_in_dense,
+            dense_prefix_depth,
+            mined_items,
+        );
 
+        let Some(mined_builder) = CanonicalTreeBuilder::restore(
```

Comment:
You need this line:
```
        let mined_items = dedup_tree_updates(mined_items);
```

somewhere here, otherwise the `mined_items` can contain duplicate items